File: TCS/Graphs/bfs_dfs.py
from collections import deque


class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.adj = [[] for _ in range(vertices)]

    def add_edge(self, v, w):
        self.adj[v].append(w)
        self.adj[w].append(v)  # Assuming an undirected graph

    def BFS(self, s):
        visited = [False] * self.V
        queue = deque([s])
        visited[s] = True

        while queue:
            node = queue.popleft()
            print(node, end=" ")

            for neighbor in self.adj[node]:
                if not visited[neighbor]:
                    queue.append(neighbor)
                    visited[neighbor] = True

    def DFSUtil(self, v, visited):
        # Mark the current node as visited and print it
        visited[v] = True
        print(v, end=" ")

        # Recur for all the vertices adjacent to this vertex
        for neighbor in self.adj[v]:
            if not visited[neighbor]:
                self.DFSUtil(neighbor, visited)

    def DFS(self, v):
        # Mark all the vertices as not visited
        visited = [False] * self.V

        # Call the recursive helper function to print DFS traversal
        self.DFSUtil(v, visited)


# Example usage
if __name__ == "__main__":
    g = Graph(6)
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 3)
    g.add_edge(2, 4)
    g.add_edge(3, 5)

    print("BFS starting from node 0:")
    g.BFS(0)

File: TCS/Graphs/bfs_dfs.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

class Graph
{
  int V;
  vector<vector<int>> adj;

public:
  Graph(int V);
  void addEdge(int v, int w);
  void DFS(int s);
  void DFSUtil(int s, vector<bool> &visited);
  void BFS(int s);
};

Graph::Graph(int V)
{
  this->V = V;
  adj.resize(V);
}

void Graph::addEdge(int v, int w)
{
  adj[v].push_back(w);
  adj[w].push_back(v); // Assuming an undirected graph
}

void Graph::DFSUtil(int s, vector<bool> &visited)
{
  visited[s] = true;
  cout << s << " ";

  for (int neighbor : adj[s])
  {
    if (!visited[neighbor])
    {
      DFSUtil(neighbor, visited);
    }
  }
}

void Graph::DFS(int s)
{
  vector<bool> visited(V, false);
  DFSUtil(s, visited);
}

void Graph::BFS(int s)
{
  vector<bool> visited(V, false);
  queue<int> q;

  visited[s] = true;
  q.push(s);

  while (!q.empty())
  {
    int node = q.front();
    q.pop();
    cout << node << " ";

    for (int neighbor : adj[node])
    {
      if (!visited[neighbor])
      {
        visited[neighbor] = true;
        q.push(neighbor);
      }
    }
  }
}

int main()
{
  Graph g(6);
  g.addEdge(0, 1);
  g.addEdge(0, 2);
  g.addEdge(1, 3);
  g.addEdge(2, 4);
  g.addEdge(3, 5);

  cout << "DFS starting from node 0:" << endl;
  g.DFS(0);

  return 0;
}
